lxd/storage/drivers/driver_btrfs.go

```diff
@@ -0,0 +1,1514 @@
+		isBlockDev = shared.IsBlockdevPath(source)
+
+		if isBlockDev {
+			output, err := MakeFSType(source, "btrfs", &MkfsOptions{Label: d.name})
+			if err != nil {
+				return fmt.Errorf("Failed to create btrfs: %v (%s)", err, output)
+			}
+		} else {
+			if isBtrfsSubvolume(source) {
+				subvols, err := btrfsSubvolumesGet(source)
+				if err != nil {
+					return fmt.Errorf("Could not determine if existing btrfs subvolume ist empty: %s", err)
+				}
+				if len(subvols) > 0 {
+					return fmt.Errorf("Requested btrfs subvolume exists but is not empty")
+				}
+			} else {
+				cleanSource := filepath.Clean(source)
+				lxdDir := shared.VarPath()
+
+				if shared.PathExists(source) && !isOnBtrfs(source) {
+					return fmt.Errorf("Existing path is neither a btrfs subvolume nor does it reside on a btrfs filesystem")
+				} else if strings.HasPrefix(cleanSource, lxdDir) {
+					if cleanSource != poolMntPoint {
+						return fmt.Errorf("btrfs subvolumes requests in LXD directory %q are only valid under %q\n(e.g. source=%s)", shared.VarPath(), shared.VarPath("storage-pools"), poolMntPoint)
+					} else if d.state.OS.BackingFS != "btrfs" {
+						return fmt.Errorf("Creation of btrfs subvolume requested but %q does not reside on btrfs filesystem", source)
+					}
+				}
+
+				err := btrfsSubvolumeCreate(source)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	var err error
+	var devUUID string
+	mountFlags, mountOptions := ResolveMountOptions(d.getMountOptions())
+	mountFlags |= d.remount
+
+	if isBlockDev {
+		devUUID, _ = shared.LookupUUIDByBlockDevPath(source)
+		// The symlink might not have been created even with the delay
+		// we granted it above. So try to call btrfs filesystem show and
+		// parse it out. (I __hate__ this!)
+		if devUUID == "" {
+			devUUID, err = btrfsLookupFsUUID(source)
+			if err != nil {
+				return err
+			}
+		}
+		d.config["source"] = devUUID
+
+		// If the symlink in /dev/disk/by-uuid hasn't been created yet
+		// aka we only detected it by parsing btrfs filesystem show, we
+		// cannot call StoragePoolMount() since it will try to do the
+		// reverse operation. So instead we shamelessly mount using the
+		// block device path at the time of pool creation.
+		err = tryMount(source, GetPoolMountPath(d.name), "btrfs", mountFlags, mountOptions)
+	} else {
+		_, err = d.Mount()
+	}
+	if err != nil {
+		return err
+	}
+
+	// Create default subvolumes.
+	subvolumes := []string{
+		filepath.Join(poolMntPoint, "containers"),
+		filepath.Join(poolMntPoint, "containers-snapshots"),
+		filepath.Join(poolMntPoint, "custom"),
+		filepath.Join(poolMntPoint, "custom-snapshots"),
+		filepath.Join(poolMntPoint, "images"),
+		filepath.Join(poolMntPoint, "virtual-machines"),
+		filepath.Join(poolMntPoint, "virtual-machines-snapshots"),
+	}
+
+	for _, subvol := range subvolumes {
+		err := btrfsSubvolumeCreate(subvol)
+		if err != nil {
+			return fmt.Errorf("Could not create btrfs subvolume: %s", subvol)
+		}
+	}
```

Comment:
I don't think we should be doing any of this anymore. It's never been clear why we did it before and as far as I'm aware, it's not really getting us any benefit feature wise, possibly just making things slower and more complicated.

Lets keep them as regular directories, which the backend will create for us.